### PR TITLE
Add rwlock.

### PR DESCRIPTION
**Pull request by BaikalHu**
Add rwlock:
Use Write-preferring.

### CODE REVIEW

File: kernel/base/ipc/los_rwlock.c [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+    if (NULL == puwRWLockHandle)
+    {
+        return LOS_ERRNO_RWLOCK_PTR_NULL;
+    }
+
+    uwIntSave = LOS_IntLock();
+    if (LOS_ListEmpty(&g_stUnusedRWLockList))
+    {
+        LOS_IntRestore(uwIntSave);
+        OS_GOTO_ERR_HANDLER(LOS_ERRNO_RWLOCK_ALL_BUSY);
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: ...

File: kernel/base/ipc/los_rwlock.c [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+    LOS_ListInit(&pstRWLockCreated->stWLockList);
+    pstRWLockCreated->ucRWLockCount = 0;
+    pstRWLockCreated->ucRWLockStat  = OS_RWLOCK_USED;
+    pstRWLockCreated->ucRCount      = 0;
+    pstRWLockCreated->pstWOwner     = (LOS_TASK_CB *)NULL;
+    *puwRWLockHandle                = (UINT32)pstRWLockCreated->ucRWLockID;
+    LOS_IntRestore(uwIntSave);
+    return LOS_OK;
+ErrHandler:
+    OS_RETURN_ERROR_P2(uwErrLine, uwErrNo);
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: ...

File: kernel/base/ipc/los_rwlock.c [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+}
+
+/*****************************************************************************
+ Function     : LOS_RWLockDelete
+ Description  : Delete a rwlock,
+ Input        : uwRWLockHandle------Mutex operation handle
+ Output       : None
+ Return       : LOS_OK on success ,or error code on failure
+ *****************************************************************************/
+LITE_OS_SEC_TEXT_INIT UINT32 LOS_RWLockDelete(UINT32 uwRWLockHandle)
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: The interrupt lock window can be reduced.

File: kernel/base/ipc/los_rwlock.c [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    LOS_TASK_CB  *pstRunTsk;
+
+    if (uwRWLockHandle >= (UINT32)LOSCFG_BASE_IPC_RWLOCK_LIMIT)
+    {
+        OS_RETURN_ERROR(LOS_ERRNO_RWLOCK_INVALID);
+    }
+
+    pstRWLockPended = GET_RWLOCK(uwRWLockHandle);
+    uwIntSave = LOS_IntLock();
+    if (OS_RWLOCK_UNUSED == pstRWLockPended->ucRWLockStat)
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: Should be checked before intlock

File: kernel/base/ipc/los_rwlock.c [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    {
+        LOS_IntRestore(uwIntSave);
+        OS_RETURN_ERROR(LOS_ERRNO_RWLOCK_INVALID);
+    }
+
+    if (OS_INT_ACTIVE)
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: Should be checked before intlock

File: kernel/base/ipc/los_rwlock.c [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    {
+        LOS_IntRestore(uwIntSave);
+        return LOS_ERRNO_RWLOCK_PEND_INTERR;
+    }
+
+    pstRunTsk = (LOS_TASK_CB *)g_stLosTask.pstRunTask;
+    if (pstRWLockPended->ucRWLockCount == 0)
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: This if and the next one can be combined as:

if (WriteCount == 0)
{
    ...
    return OK;
}

File: kernel/base/ipc/los_rwlock.c [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+
+/*****************************************************************************
+ Function     : LOS_WLock
+ Description  : Specify the rwlock w lock operation,
+ Input        : uwRWLockHandle ------ RWLock operation handleone,
+                uwTimeOut  ------- waiting time,
+ Output       : None
+ Return       : LOS_OK on success ,or error code on failure
+ *****************************************************************************/
+LITE_OS_SEC_TEXT UINT32 LOS_WLock(UINT32 uwRWLockHandle, UINT32 uwTimeout)
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: Checks in intlock can be reduced.

File: kernel/base/ipc/los_rwlock.c [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    {
+        pstRWLockPended->ucRWLockCount++;
+        pstRWLockPended->ucRCount++;
+        LOS_IntRestore(uwIntSave);
+        return LOS_OK;
+    }
+
+    if (pstRWLockPended->ucRWLockCount == pstRWLockPended->ucRCount)
+    {
+        pstRWLockPended->ucRWLockCount++;
```
Comment (chaifangming) [inline kernel/base/ipc/los_rwlock.c]: Seems just WLockCount and RLockCount is simpler, needless define a RWCount.

File: kernel/base/include/los_rwlock.h [hunk 9]
```diff
@@ -0,0 +1,377 @@
+/*----------------------------------------------------------------------------
+ * Copyright (c) <2013-2015>, <Huawei Technologies Co., Ltd>
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: maintain <2013-2015> or change every files.

File: kernel/base/include/los_rwlock.h [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+ * @ingroup los_rwlock
+ * RWLock object.
+ */
+typedef struct
+{
+    UINT8           ucRWLockStat;       /**< State OS_RWLOCK_UNUSED,OS_RWLOCK_USED  */
+    UINT8           ucRCount;
+    UINT8           ucRWLockCount;      /**< Times of locking a RWLock */
+    UINT32          ucRWLockID;         /**< Handle ID*/
+    LOS_DL_LIST     stRLockList;       /**< RLock linked list*/
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: change to Pend.

File: kernel/base/include/los_rwlock.h [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+ *@retval #LOS_ERRNO_RWLOCK_PEND_INTERR        The rwlock is being locked during an interrupt.
+ *@retval #LOS_ERRNO_RWLOCK_PEND_IN_LOCK       The rwlock is waited on when the task scheduling is disabled.
+ *@retval #LOS_ERRNO_RWLOCK_TIMEOUT            The rwlock waiting times out.
+ *@retval #LOS_OK                              The rwlock is successfully locked.
+ *@par Dependency:
+ *<ul><li>los_rwlock.h: the header file that contains the API declaration.</li></ul>
+ *@see LOS_RWLockCreate | LOS_RUnLock
+ *@since Huawei LiteOS
+ */
+UINT32 LOS_RLock(UINT32 uwRWLockHandle, UINT32 uwTimeout);
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: LOS_RWLockRdLock or LOS_RWReadLock? Which is better?

File: kernel/base/include/los_rwlock.h [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+ *@retval #LOS_ERRNO_RWLOCK_PEND_INTERR        The rwlock is being locked during an interrupt.
+ *@retval #LOS_ERRNO_RWLOCK_PEND_IN_LOCK       The rwlock is waited on when the task scheduling is disabled.
+ *@retval #LOS_ERRNO_RWLOCK_TIMEOUT            The rwlock waiting times out.
+ *@retval #LOS_OK                              The rwlock is successfully locked.
+ *@par Dependency:
+ *<ul><li>los_rwlock.h: the header file that contains the API declaration.</li></ul>
+ *@see LOS_RWLockCreate | LOS_WUnLock
+ *@since Huawei LiteOS
+ */
+UINT32 LOS_WLock(UINT32 uwRWLockHandle, UINT32 uwTimeout);
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: The same.

File: kernel/base/include/los_rwlock.h [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+ *@param uwRWLockHandle    [IN] Handle of the rwlock to be unlock. The value of handle should be in [0, LOSCFG_BASE_IPC_RWLOCK_LIMIT - 1].
+ *
+ *@retval #LOS_ERRNO_RWLOCK_INVALID            The rwlock state (for example, the rwlock does not exist or is not in use or owned by other thread) is not applicable for the current operation.
+ *@retval #LOS_OK                              The rwlock is successfully unlock.
+ *@par Dependency:
+ *<ul><li>los_rwlock.h: the header file that contains the API declaration.</li></ul>
+ *@see LOS_RWLockCreate | LOS_RLock
+ *@since Huawei LiteOS
+ */
+UINT32 LOS_RUnLock(UINT32 uwRWLockHandle);
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: The same.

File: kernel/base/include/los_rwlock.h [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+ *@retval #LOS_ERRNO_RWLOCK_INVALID            The rwlock state (for example, the rwlock does not exist or is not in use or owned by other thread) is not applicable for the current operation.
+ *@retval #LOS_OK                              The rwlock is successfully unlock.
+ *@par Dependency:
+ *<ul><li>los_rwlock.h: the header file that contains the API declaration.</li></ul>
+ *@see LOS_RWLockCreate | LOS_WLock
+ *@since Huawei LiteOS
+ */
+UINT32 LOS_WUnLock(UINT32 uwRWLockHandle);
+
+
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: The same.

File: kernel/base/ipc/los_rwlock.c [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+
+    g_pstAllRWLock = (RWLOCK_CB_S *)LOS_MemAlloc(m_aucSysMem0, (LOSCFG_BASE_IPC_RWLOCK_LIMIT * sizeof(RWLOCK_CB_S)));
+    if (NULL == g_pstAllRWLock)
+    {
+        return LOS_ERRNO_RWLOCK_NO_MEMORY;
+    }
+
+    for (uwIndex = 0; uwIndex < LOSCFG_BASE_IPC_RWLOCK_LIMIT; uwIndex++)
+    {
+        pstRWLockNode                 = ((RWLOCK_CB_S *)g_pstAllRWLock) + uwIndex;
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: ok

File: kernel/base/ipc/los_rwlock.c [hunk 16]
```diff
@@ -0,0 +1,499 @@
...
+
+    uwIntSave = LOS_IntLock();
+    if (LOS_ListEmpty(&g_stUnusedRWLockList))
+    {
+        LOS_IntRestore(uwIntSave);
+        OS_GOTO_ERR_HANDLER(LOS_ERRNO_RWLOCK_ALL_BUSY);
+    }
+
+    pstUnusedRWLock                 = LOS_DL_LIST_FIRST(&(g_stUnusedRWLockList));
+    LOS_ListDelete(pstUnusedRWLock);
```
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: ok. It reduce the time of interrupt lock time and still safe.

[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: Will sem and mutex do the similar change?

File: kernel/base/ipc/los_rwlock.c (hunk 1 above)
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: Do you mean the label?

File: kernel/base/ipc/los_rwlock.c (hunk 2 above)
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: Do you mean the label?

File: kernel/base/ipc/los_rwlock.c (hunk 3 above)
Comment (BaikalHu) [reply kernel/base/ipc/los_rwlock.c]: Can it be reduced?

File: kernel/base/include/los_rwlock.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (BaikalHu) [reply kernel/base/include/los_rwlock.h]: chose LOS_RWReadLock

**Pull request closed without merging**